python/ray/air/execution/actor_manager.py

```diff
@@ -0,0 +1,244 @@
+        """
+        raise NotImplementedError
+
+    def get_actor_resources(self, actor_info: ActorInfo) -> Optional[AcquiredResources]:
+        """Returns the acquired resources of an actor that has been started.
+
+        This will return ``None`` if the actor has not been started, yet.
+
+        Args:
+            actor_info: Actor info object of the actor to get resources for.
+        """
+        raise NotImplementedError
+
+    def schedule_task(
+        self,
+        actor_info: ActorInfo,
+        task_spec: Tuple[str, Iterable[Any], Dict[str, Any]],
+        result_cls: Type[FutureResult],
+    ) -> None:
+        """Track (asynchronous) future associated to a tracked actor.
+
+        This method will schedule the remote task and track its execution. Once the
+        future is resolved, it will be emitted as an event. The event
+        class will be ``result_cls``, which should extend ``FutureResult``.
+
+        The ``task_spec`` argument is submitted as a tuple of
+        ``(method_name, args, kwargs)``. This will result in the future
+        ``future = actor.method_name.remote(*args, **kwargs)`` to be scheduled on
+        the actor and tracked in the manager.
+
+        Events:
+            ``result_cls``: Emitted once the future resolved.
+            ``FutureFailed``: Emitted if the future resolution fails due to a
+                ``RayTaskError`, indicating method execution ran into an exception.
+            ``ActorFailed``: Emitted if the future resolution fails due to a
+                ``RayActorError``, indicating the actor died.
+
+        Args:
+            actor_info: Actor info object to track future for.
+            task_spec: Tuple of ``(method_name, args, kwargs)`` to be scheduled as
+                task futures on the actor.
+            result_cls: Class extending ``FutureResult`` to emit when the future
+                resolves successfully.
+
+        """
+        raise NotImplementedError
+
+    def schedule_sync_tasks(
+        self,
+        actors_tasks: Dict[ActorInfo, Tuple[str, Iterable[Any], Dict[str, Any]]],
+        result_cls: Type[FutureResult],
+    ) -> None:
+        """Track synchronous futures associated to multiple tracked actors.
+
+        This method will schedule remote tasks and track their execution
+        synchronously on all provided actors. This means the multi result event will
+        only be emitted once _all_ provided tasks resolved.
+
+        Once all futures resolved successfully, a ``MultiFutureResult`` event
+        containing all sub results will be emitted as an event.
+        The sub results will be of type ``result_cls``, which should
+        extend ``FutureResult``.
+
+        If any of the tasks fail, a ``MultiFutureFailed`` event will be emitted.
+        The ``exception`` parameter will contain the first observed exception.
+        Unfinished tasks will be cancelled. The sub results will be of type
```

Comment:
why cancel the unfinished tasks?
the contract current rllib actor manager has with the rest of the codebase is return everything that succeeded, return error for anything that failed.
also, does FutureResult contain the actual data or just the object_ref?

Reply:
FutureResult should contain the actual data (I can make this more clear in the docs).

The reason we cancel the remaining tasks is because these are _synchronous_ futures. E.g. in Ray Train if we schedule training of episode X on N actors, we only care about the results if we can get all N results.

Can you elaborate more on how rllib does this? Are these actually synchronous futures that should all return at the same time? My understanding was that these are unsynchronous in which case the behavior is as described.